### PR TITLE
Adding continued_pretraining task

### PR DESCRIPTION
**Pull request by BramVanroy**
As discussed internally, it could be very valuable to expand the alignment handbook with a prior step to sft/dpo, namely continued pretraining. This is especially useful when doing domain or language adaptation of a language model.

The main contribution is the addition of the `run_continued_pretraining.py` script, as suggested by @lewtun. (A shorter alternative could be `cpt`, although that is not a well-known abbreviation.) It is very similar to the `run_sft.py` script with only minor changes to ensure that no chat templating is needed. We can even re-use the `SFTConfig` class for setting hyperparameters (especially sequence length). I've also ensured that the eval dataset is allowed to be optional but that a training set must always be present.

**One question**: I am not sure about how the `SFTTrainer` deals with truncation. My assumption is that the code now will simply truncate long samples but we'd likely want something that does not throw so much data away, like with the `run_clm.py` script over at transformers, where the data is first tokenized (no truncation) and then [groupe](https://github.com/huggingface/transformers/blob/87a0783dde7291876d096d9dba5685dc1b7bd4c6/examples/pytorch/language-modeling/run_clm.py#L526)d in `block_size` items. I did not currently implement this because I wasn't sure what the best approach would be, considering the SFTTrainer does the packing and such under the hood, which amy already be covering this?

There are also two changes that are related to the existing scripts:

- model_utils.get_tokenizer now has an `auto_set_chat_template` (default: True). In SFT/DPO we always want to make sure that the tokenizer has a `chat_template` set, so this function used to automatically set a default template if the tokenizer did not have one yet. However, during continued pretraining we do not want to set that chat template yet, so this automatic step should be skippable. When running the `run_continued_pretraining.py` script, `get_tokenizer` will have `auto_set_chat_template=False` for this reason. Since `True` is the default value, nothing changes for the sft/dpo scripts.
- the option to specify dataset_configs is added to all scripts. With a new `data_arguments.dataset_configs`, a user can specify which config to use for which dataset. If not given, will use `None` but otherwise the argument must be the same length as the number of datasets that are given in the dataset mixer. So if you specify to use two datasets, your `dataset_configs` must also contain two items, one for each dataset. (Or it can of course be left out.)

An example recipe is added for language adapting `gpt2` to Dutch. I ran the scripts so I can verify that they work, but I did not do any hyperparameter optimisation and the output model is quite bad (it's `gpt2-small` after all). But it serves as a guideline for users.

### CODE REVIEW

**Comment by BramVanroy** [comment]
I'm not sure that I understand the point of this failing test or why it is failing (since the default case remains unchanged) so advice is welcome.

File: README.md [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 # The Alignment Handbook
 
-Robust recipes to align language models with human and AI preferences.
+Robust recipes to continue pretraining and to align language models with human and AI preferences.
```
Comment (lewtun) [inline README.md:11]: Cool!

File: README.md [hunk 2]
```diff
@@ -32,8 +32,8 @@ The Alignment Handbook aims to fill that gap by providing the community with a s
 
 This project is simple by design and mostly consists of:
 
-* [`scripts`](./scripts/) to train and evaluate chat models. Each script supports distributed training of the full model weights with DeepSpeed ZeRO-3, or LoRA/QLoRA for parameter-efficient fine-tuning.
-* [`recipes`](./recipes/) to reproduce models like Zephyr 7B. Each recipe takes the form of a YAML file which contains all the parameters associated with a single training run.
+* [`scripts`](./scripts/) to train and evaluate models. Three steps are included: continued pretraining, supervised-finetuning (sft) for chat, and reinforcement leanring with DPO. Each script supports distributed training of the full model weights with DeepSpeed ZeRO-3, or LoRA/QLoRA for parameter-efficient fine-tuning.
```
Comment (lewtun) [inline README.md:35]: nit: DPO is technically an RL-free algorithm :)

```suggestion
* [`scripts`](./scripts/) to train and evaluate models. Three steps are included: continued pretraining, supervised-finetuning (SFT) for chat, and preference alignment with DPO. Each script supports distributed training of the full model weights with DeepSpeed ZeRO-3, or LoRA/QLoRA for parameter-efficient fine-tuning.
```

File: README.md [hunk 3]
```diff
@@ -32,8 +32,8 @@ The Alignment Handbook aims to fill that gap by providing the community with a s
...
+* [`recipes`](./recipes/) to reproduce models like Zephyr 7B. Each recipe takes the form of a YAML file which contains all the parameters associated with a single training run. A `gpt2-nl` recipe is also given to illustrate how this handbook can be used for language or domain adaptation, e.g. by continuing to pretrain on a different language, and then sft and dpo tuning the result.
```
Comment (lewtun) [inline README.md:36]: turbo nit:
```suggestion
* [`recipes`](./recipes/) to reproduce models like Zephyr 7B. Each recipe takes the form of a YAML file which contains all the parameters associated with a single training run. A `gpt2-nl` recipe is also given to illustrate how this handbook can be used for language or domain adaptation, e.g. by continuing to pretrain on a different language, and then SFT and DPO tuning the result. 
```

File: recipes/gpt2-nl/README.md
Comment (lewtun) [inline recipes/gpt2-nl/README.md:1]: WDYT about calling this recipe `continued-pretraining` to make it clear to the repo user what it's about?

File: recipes/gpt2-nl/README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
+# Language Adaptation through Continued Pretraining
+
+This directory shows a base example of how to use continued pretraining and further tuning to adapt a language model to new data (e.g. a new language or domain).
+
+Three steps are needed: continued pretraining (`continued_pretraining`), supervised finetuning (`sft`), and direct preference optimisation (`dpo`). In this dummy example we'll continue pretraining gpt2 on Dutch raw data, then sft-tuning it, and finally aligning it with DPO. Note that no extensive hyperparameters were tested in this example and that the output models are bad - it is just to show you how you can use the scripts for LM adaptation. The scripts work on 4x 3090s (24GB VRAM). If you have less powerful hardware you may need to reduce the batch size.
```
Comment (lewtun) [inline recipes/gpt2-nl/README.md:5]: I agree with you that we can shorten this term to just CPT to harmonise with the other approaches and save a few characters:
```suggestion
Three steps are needed: continued pretraining (`cpt`), supervised finetuning (`sft`), and direct preference optimisation (`dpo`). In this dummy example we'll continue pretraining gpt2 on Dutch raw data, then sft-tuning it, and finally aligning it with DPO. Note that no extensive hyperparameters were tested in this example and that the output models are bad - it is just to show you how you can use the scripts for LM adaptation. The scripts work on 4x 3090s (24GB VRAM). If you have less powerful hardware you may need to reduce the batch size.
```

File: scripts/run_continued_pretraining.py
Comment (lewtun) [inline scripts/run_continued_pretraining.py:1]: As mentioned above, I agree we can call this `run_cpt.py`

**Review by lewtun — APPROVED** [review]
Thanks for this epic PR @BramVanroy - super cool to see how flexible the `SFTTrainer` is at both pretraining and fine-tuning 🤯 !

All the changes LGTM and I left a few nits about naming things. Regarding your question on truncation, when packing is enabled in the `SFTTrainer` then no truncation is applied so I think we should be good there

File: recipes/gpt2-nl/README.md
Comment (BramVanroy) [reply recipes/gpt2-nl/README.md:1]: Done!

**Comment by lewtun** [comment]
Thanks for iterating! Excited to see what the community does with this for domains like code and math 🔥

**Pull request merged**